### PR TITLE
Add CBQRI support for supervisor domains

### PR DESCRIPTION
**Pull request by ved-rivos**

### CODE REVIEW

File: chapter9.adoc [hunk 1]
```diff
@@ -0,0 +1,334 @@
...
+
+When the `srmcfg` CSR is read at privilege modes less than M, the value returned
+for the `RCID` and `MCID` fields of the register is computed using <<eq-1>>.
+
+[latexmath#eq-1,reftext="equation ({counter:eqs0})"]
+++++
+\begin{equation}
+RCID-value = (SSM == 1) ? (srmcfg.RCID & ((1 << SRL) - 1)) : srmcfg.RCID
+MCID-value = (SSM == 1) ? (srmcfg.MCID & ((1 << SML) - 1)) : srmcfg.MCID
+\end{equation}
```
Comment (rsahita) [inline chapter9.adoc:131]: please split these into separate equations else they render on the same line (in very very tiny font)

File: chapter9.adoc [hunk 2]
```diff
@@ -0,0 +1,334 @@
...
+endif
+if SSM == 1
+    srmcfg.MCID = (srmcfg.MCID & ~((1 << SML) - 1)) | (MCID-value & ((1 << SML) - 1))
+else
+    if ((MCID-value & ~((1 << SML) - 1)) | ((1 << SML) - 1)) != MAX_MCID_VALUE
+        srmcfg.MCID = MCID-value
+    endif
+endif
+\end{equation}
+++++
```
Comment (rsahita) [inline chapter9.adoc:156]: please split these into separate equations else they render on the same line (in very very tiny font).
also would be great to make sure they render as code blocks (o/w it ends up on a single line).

**Review by rsahita — COMMENTED** [review]
some minor changes needed

File: chapter9.adoc (hunk 1 above)
Comment (ved-rivos) [reply chapter9.adoc]: Fixed

File: chapter9.adoc (hunk 2 above)
Comment (ved-rivos) [reply chapter9.adoc]: Fixed

File: chapter9.adoc [hunk 3]
```diff
@@ -0,0 +1,346 @@
...
+----
+RCID-value = (SSM == 1) ? (srmcfg.RCID & ((1 << SRL) - 1)) : srmcfg.RCID
+MCID-value = (SSM == 1) ? (srmcfg.MCID & ((1 << SML) - 1)) : srmcfg.MCID
+----
+
+<<<
+
+On a write to the `srmcfg` CSR at privilege modes less than M, the value
+stored in the `RCID` and `MCID` fields of the register is computed using
+<<eq-3>> and <<eq-4>> respetively.
```
Comment (rsahita) [inline chapter9.adoc:137]: are the eq-3 and eq-4 references valid?

File: chapter9.adoc [hunk 4]
```diff
@@ -0,0 +1,346 @@
...
+
+.`RCID` and `MCID` read value processing
```
Comment (rsahita) [inline chapter9.adoc:139]: should this be RCID and MCID write value processing?

Comment (ved-rivos) [reply chapter9.adoc]: fixed.

**Review by rsahita — APPROVED** [review]

File: chapter9.adoc [hunk 5]
```diff
@@ -0,0 +1,346 @@
...
+
+The Ssqosid extension cite:[SSQOSID] provides a read/write S/HS-mode register
+(`srmcfg`) to configure QoS Identifiers to be used with requests made by the
+hart to shared resources. The CBQRI specification specifies the IOMMU extension
+to associate QoS Identifiers with requests made by devices to shared resources.
+
+With supervisor domains, the resource accesses from a SD or the RDSM must not
+be observable using the resource usage monitors to entities that are not in
+their TCB.  The resource allocations for a SD or the RDSM must not be affected
+by entities that are not in their TCB.
```
Comment (rsahita) [inline chapter9.adoc:27]: move security requirements up to the front to explain the rationale for the sub-extension

File: chapter9.adoc [hunk 6]
```diff
@@ -0,0 +1,346 @@
...
+  {bits:  1, name: 'SSM'},
+  {bits:  1, name: 'SDEDBGALW'},
+  {bits:  4, name: 'SRL'},
+  {bits:  4, name: 'SML'},
+  {bits:  4, name: 'SQRID'},
+  {bits:  4, name: 'MQRID'},
+  {bits:  8, name: 'WPRI'},
+], config:{lanes: 4, hspace:1024}}
+....
+
```
Comment (eckhard-delfs-qualcomm) [inline chapter9.adoc:122]: The following sentences may help to understand the rationale of the subsequent logic equations:

_The srmcfg CSR is provided with a privilege mode dependent read and write logic which enables the RDSM to configure supervisor domain specific prefixes for RCID and MCID fields. These prefixes remain invisible to supervisor domains which can only configure RCID and MCID suffixes within the boundaries set by RDSM in msdfg._

Comment (ved-rivos) [reply chapter9.adoc:122]: Thanks. That is a good addition. I will update the PR to include this.

File: chapter9.adoc (hunk 5 above)
Comment (ved-rivos) [reply chapter9.adoc:27]: I think this is in a  logical place. The paragraph 1 and 2 and 3 provide an introduction to what CBQRI is. This parapgraph - paragraph 4 - explains why this extension is needed. The para 1, 2, and 3 are needed to appreciate para 4.

Comment (rsahita) [reply chapter9.adoc:27]: It came up as a question - at least a heading such as “security objectives “ would be nice to add to avoid readers wondering

Comment (ved-rivos) [reply chapter9.adoc]: Please see updated.

File: chapter9.adoc (hunk 6 above)
Comment (ved-rivos) [reply chapter9.adoc:141]: Please see the updated text.

File: chapter9.adoc [hunk 7]
```diff
@@ -0,0 +1,372 @@
...
+The `srmcfg` CSR is provided with a privilege mode dependent read and write
+logic which enables the RDSM to configure supervisor domain specific prefixes for
+`RCID` and `MCID` fields. These prefixes remain invisible to supervisor domains
+which can only configure `RCID` and `MCID` suffixes within the boundaries set by
+RDSM in `msdfg`.
```
Comment (rsahita) [inline chapter9.adoc:146]:
```suggestion
RDSM in `msdcfg`.
```

**Review by rsahita — APPROVED** [review]

File: chapter9.adoc (hunk 7 above)
Comment (eckhard-delfs-qualcomm) [reply chapter9.adoc:146]: Thanks @ved-rivos. 
Beyond the typo identified by Ravi (my fault) looks good to me.

File: chapter9.adoc [hunk 8]
```diff
@@ -0,0 +1,372 @@
...
+to selecting a `RCID` and `MCID` in `srmcfg` from the allocated range. Although
+the RDSM could limit access to `srmcfg` to prevent direct domain configuration,
+enabling domains to program `srmcfg` directly is preferred for performance
+reasons. To facilitate this, the Smqosid extension introduces `SRL`
+(supervisor-RCID-length) and `SML` (supervisor-MCID-length) fields in the
+`msdcfg` register to configure the usable range of `RCID` and `MCID`. These
+fields are WARL, with legal values from 0 to 12. Furthermore, a
+supervisor-srmcfg-mode (`SSM`) field is introduced, with legal values of 0 and
+1. These settings, along with `SRL` and `SML`, determine the range of `RCID` and
+`MCID` values an SD can program into the `srmcfg` register.
```
Comment (rsahita) [inline chapter9.adoc:124]: clarify requirement is for these ranges to be disjoint

**Review by rsahita — CHANGES_REQUESTED** [review]
some minor changes needed to convert equations to if,then,else and other descriptive updates.

File: chapter9.adoc (hunk 7 above)
Comment (ved-rivos) [reply chapter9.adoc:146]: fixed

File: chapter9.adoc (hunk 8 above)
Comment (ved-rivos) [reply chapter9.adoc]: done - in new PR

**Pull request closed without merging**